### PR TITLE
Allow factory.get() to create from a ISO calendar tuple

### PR DESCRIPTION
**Pull request by systemcatch**
ref #189 

**TODO**

- [x] Error handling.
- [x] Test edge cases.
- [x] Docstring.
- [x] Docs

### CODE REVIEW

File: arrow/util.py [hunk 1]
```diff
@@ -1,7 +1,7 @@
 # -*- coding: utf-8 -*-
 from __future__ import absolute_import
 
-from datetime import datetime
+import datetime
```
Comment (jadchaar) [inline arrow/util.py]: Any reason for this?

File: arrow/factory.py [hunk 2]
```diff
@@ -99,6 +99,16 @@ def get(self, *args, **kwargs):
             >>> arrow.get(date(2013, 5, 5))
             <Arrow [2013-05-05T00:00:00+00:00]>
 
+        **One** time.struct time::
+
+            >>> arrow.get(gmtime(0))
+            <Arrow [1970-01-01T00:00:00+00:00]>
+
+        **One** iso calendar ``tuple``, to get that week date in UTC::
```
Comment (jadchaar) [inline arrow/factory.py]: Let's capitalize ISO :)

File: arrow/factory.py [hunk 3]
```diff
@@ -192,6 +197,11 @@ def get(self, *args, **kwargs):
             elif isinstance(arg, struct_time):
                 return self.type.utcfromtimestamp(calendar.timegm(arg))
 
+            # (iso calendar) -> convert then from date
+            elif isinstance(arg, tuple):
```
Comment (jadchaar) [inline arrow/factory.py]: I am a bit worried that `tuple` is too broad. I wonder if there is any other datetime information we can absorb in the `get()` method that would be of type `tuple`.

File: arrow/util.py [hunk 4]
```diff
@@ -12,12 +12,30 @@ def is_timestamp(value):
...
         return True
     except TypeError:
         return False
 
 
+# Credit to https://stackoverflow.com/a/1700069
+def iso_to_gregorian(iso_year, iso_week, iso_day):
+    """Converts an iso weekday tuple into a datetime object."""
+
+    if iso_week == 0 or iso_day == 0:
```
Comment (jadchaar) [inline arrow/util.py]: We should probably split this into two conditionals so we can make the error messages clearer.

File: tests/util_tests.py [hunk 5]
```diff
@@ -29,3 +29,10 @@ def test_is_timestamp(self):
         overflow_timestamp_int = int(overflow_timestamp_float)
         with self.assertRaises((OverflowError, ValueError)):
             util.is_timestamp(overflow_timestamp_int)
+
+    def test_iso_gregorian(self):
```
Comment (jadchaar) [inline tests/util_tests.py]: We should probably test more edge cases such as those listed here: https://stackoverflow.com/q/304256/.

File: arrow/util.py (hunk 1 above)
Comment (systemcatch) [reply arrow/util.py]: Mainly to make the imports easier.

**Comment by systemcatch** [comment]
Hey @jadchaar I made most of the additions you suggested.

>I am a bit worried that `tuple` is too broad. I wonder if there is any other datetime information we can absorb in the `get()` method that would be of type `tuple`.

Looking through the `get()` method nothing else would be a `tuple`, `struct_time` is a subclass of `tuple` though so ordering does matter in the method. Maybe I could add a check like this to avoid any subclasses accidentally getting passed in the future.

```python
if issubclass(arg, tuple):
    # pass or raise here
```

File: arrow/util.py [hunk 6]
```diff
@@ -12,12 +12,31 @@ def is_timestamp(value):
...
     except TypeError:
         return False
 
 
+# Credit to https://stackoverflow.com/a/1700069
+def iso_to_gregorian(iso_year, iso_week, iso_day):
+    """Converts an iso weekday tuple into a datetime object."""
+
+    if not 1 <= iso_week <= 53:
+        raise ValueError("Iso Calendar week value must be between 1-53.")
```
Comment (jadchaar) [inline arrow/util.py]: Let's make all the casings of ISO linear in the docstrings and error messages. We should probably capitalize it fully as ISO.

File: tests/factory_tests.py [hunk 7]
```diff
@@ -140,6 +140,34 @@ def test_one_arg_iso_str(self):
 
         assertDtEqual(self.factory.get(dt.isoformat()), dt.replace(tzinfo=tz.tzutc()))
 
+    def test_one_arg_iso_calendar(self):
```
Comment (jadchaar) [inline tests/factory_tests.py]: Solid tests, but we should also add the two mentioned in the Python docs: https://docs.python.org/3/library/datetime.html#datetime.date.isocalendar

File: arrow/util.py [hunk 8]
```diff
@@ -12,12 +12,31 @@ def is_timestamp(value):
...
-        datetime.fromtimestamp(value)
+        datetime.datetime.fromtimestamp(value)
         return True
     except TypeError:
         return False
 
 
+# Credit to https://stackoverflow.com/a/1700069
+def iso_to_gregorian(iso_year, iso_week, iso_day):
+    """Converts an iso weekday tuple into a datetime object."""
```
Comment (jadchaar) [inline arrow/util.py]: Let's capitalize ISO in all the docstrings and error messages for consistency :)

File: arrow/factory.py (hunk 3 above)
Comment (jadchaar) [inline arrow/factory.py]: Seems like the best resolution to this is to expand this case to: `elif isinstance(arg, tuple) and len(tuple) == 3:` since ISO calendar is a 3-tuple.

**Comment by jadchaar** [comment]
> Hey @jadchaar I made most of the additions you suggested.
> 
> > I am a bit worried that `tuple` is too broad. I wonder if there is any other datetime information we can absorb in the `get()` method that would be of type `tuple`.
> 
> Looking through the `get()` method nothing else would be a `tuple`, `struct_time` is a subclass of `tuple` though so ordering does matter in the method. Maybe I could add a check like this to avoid any subclasses accidentally getting passed in the future.
> 
> ```python
> if issubclass(arg, tuple):
>     # pass or raise here
> ```

I think since ISO calendar is a 3-tuple, it may be substantial enough to add a check for the length of the tuple like I showed in the comments above ^. If we add any tuples in the future, we can definitely investigate this further, but I think the length check is a good starting point.

**Comment by systemcatch** [comment]
@jadchaar all changes now implemented!

File: arrow/factory.py [hunk 9]
```diff
@@ -192,6 +197,11 @@ def get(self, *args, **kwargs):
             elif isinstance(arg, struct_time):
                 return self.type.utcfromtimestamp(calendar.timegm(arg))
 
+            # (iso calendar) -> convert then from date
```
Comment (jadchaar) [inline arrow/factory.py]: Nit-picky detail, but what do you mean by "convert then"?

Comment (jadchaar) [reply arrow/factory.py]: Never mind, I misread :).
